### PR TITLE
snapshot for rewards token distribution percentage

### PR DESCRIPTION
**Pull request by Suyog007**
- created function for storing rewards token distribution percentage
- getter function added for finding the distribution pecentage in a specific day
- minor changes

### CODE REVIEW

File: core_contracts/governance/governance.py [hunk 1]
```diff
@@ -544,6 +544,16 @@ def addNewDataSource(self, _data_source_name: str, _contract_address: Address) -
         rewards = self.create_interface_score(self.addresses['rewards'], RewardsInterface)
         rewards.addNewDataSource(_data_source_name, _contract_address)
 
+    @external
+    @only_owner
+    def removeDataSource(self, _data_source_name: str, _contract_address: Address) -> None:
+        """
+        Add a new data source to receive BALN tokens. Starts with a default of
```
Comment (dbrehmer) [inline core_contracts/governance/governance.py]: Docstring is still the one for adding a data source.

File: core_contracts/rewards/rewards.py [hunk 2]
```diff
@@ -195,9 +201,34 @@ def addNewDataSource(self, _name: str, _address: Address) -> None:
...
+        """
+        Sources for data on which to base incentive rewards are removed with this
+        method.
+
+        :param _name: Identifying name for the data source.
+        :type _name: str
+        :param _address: Address of the data source.
+        :type _address: :class:`iconservice.base.address.Address`
+        """
+        if _name not in self._recipients:
```
Comment (dbrehmer) [inline core_contracts/rewards/rewards.py:220]: Needs a check that the dist percentage for this source has been set to zero. Can't remove a source unless it has zero distribution.

File: core_contracts/rewards/rewards.py [hunk 3]
```diff
@@ -195,9 +201,34 @@ def addNewDataSource(self, _name: str, _address: Address) -> None:
...
+            return
+        if not _address.is_contract:
```
Comment (dbrehmer) [inline core_contracts/rewards/rewards.py]: Does not really matter whether the source is a contract. Can skip this check.

**Review by dbrehmer — CHANGES_REQUESTED** [review]
A couple changes requested. Review is not finished so there may be more.

File: core_contracts/rewards/rewards.py (hunk 3 above)
Comment (Suyog007) [reply core_contracts/rewards/rewards.py]: sure .

File: core_contracts/rewards/rewards.py (hunk 2 above)
Comment (Suyog007) [reply core_contracts/rewards/rewards.py:220]: yeah this function is only called by owner so, assuming dist percentage will be updated at first before removing the data source. 
I will add the check for now.

File: core_contracts/rewards/RewardData.py [hunk 4]
```diff
@@ -151,3 +151,14 @@ def new_source(self, _name: str, _address: Address) -> None:
         source.name.set(_name)
         source.day.set(self._rewards._get_day())
         source.contract_address.set(_address)
+
+    def remove_source(self, _name: str, _address: Address) -> None:
+        source = self.__getitem__(_name)
+        source.name.set("")
```
Comment (ajaya-man) [inline core_contracts/rewards/RewardData.py:157]: remove

File: core_contracts/rewards/RewardData.py [hunk 5]
```diff
@@ -151,3 +151,14 @@ def new_source(self, _name: str, _address: Address) -> None:
...
+        source.day.set(0)
```
Comment (ajaya-man) [inline core_contracts/rewards/RewardData.py]: Use vardb.remove instead of vardb.set to default value

File: core_contracts/rewards/RewardData.py [hunk 6]
```diff
@@ -151,3 +151,14 @@ def new_source(self, _name: str, _address: Address) -> None:
         source.name.set(_name)
         source.day.set(self._rewards._get_day())
         source.contract_address.set(_address)
+
+    def remove_source(self, _name: str, _address: Address) -> None:
```
Comment (ajaya-man) [inline core_contracts/rewards/RewardData.py]: _address parameter is not being used.

File: core_contracts/rewards/RewardData.py [hunk 7]
```diff
@@ -151,3 +151,14 @@ def new_source(self, _name: str, _address: Address) -> None:
...
+        source = self.__getitem__(_name)
```
Comment (ajaya-man) [inline core_contracts/rewards/RewardData.py:158]: Check if _name exists in self._names

File: core_contracts/rewards/rewards.py [hunk 8]
```diff
@@ -71,6 +74,8 @@ def on_update(self) -> None:
         super().on_update()
         self._data_source_db['sICX/bnUSD'].precomp.set(0)
         self._data_source_db['sICX/bnUSD'].offset.set(0)
+        for recipient in self._recipients:
```
Comment (ajaya-man) [inline core_contracts/rewards/rewards.py:78]: Change it to a score update method and remove the previous update codes.

File: core_contracts/rewards/RewardData.py (hunk 7 above)
Comment (Suyog007) [reply core_contracts/rewards/RewardData.py:158]: used not in to check. 
let me know if this will create trouble.

**Review by ajaya-man — APPROVED** [review]
Looks Good to me.

**Review by dbrehmer — APPROVED** [review]
Looks good.

**Pull request merged**